utils/estimated_histogram.hh

```diff
@@ -47,13 +47,376 @@
+    static constexpr uint8_t BASESHIFT = (MIN == 0) ? 0 : pow2_rank_constexpr(MIN);
+    static constexpr uint64_t LOWER_BITS_MASK = (1 << RESOLUTION_BITS) - 1;
+
+    std::vector<uint64_t> buckets;
+
+    pseudo_floatingpoint_histogram() {
+        assert(BASESHIFT >= RESOLUTION);
+        buckets.resize(NUM_BUCKETS, 0);
+    }
+
+    /*!
+     * \brief Returns the bucket limit given the bucket id.
+     * The first and last bucket will always return the MIN and MAX respectively.
+     *
+     */
+    uint64_t get_bucket_limit(uint16_t bucket_id) const {
+        if (bucket_id == 0) {
+            return MIN;
+        }
+        if (bucket_id == NUM_BUCKETS - 1) {
+            return MAX;
+        }
+        int16_t range_id = (bucket_id >> RESOLUTION_BITS);
+        return (1 << (range_id + BASESHIFT)) +  ((bucket_id & LOWER_BITS_MASK) << (range_id + BASESHIFT - RESOLUTION_BITS));
+    }
+
+    /*!
+     * \brief Find the bucket index for a given value
+     * The position of a value that is lower or equal to Min will always be 0.
+     * The position of a value is that is higher or equal to MAX will always be NUM_BUCKETS - 1.
+     */
+    uint16_t find_bucket_pos(uint64_t val) const {
+        if (val >= MAX) {
+            return NUM_BUCKETS - 1;
+        }
+        if (val <= MIN) {
+            return 0;
+        }
+        uint16_t range_id = pow2_rank(val);
+        val >>= range_id - RESOLUTION_BITS; // leave the top most N+1 bits where N is the resolution.
+        return ((range_id - BASESHIFT) << RESOLUTION_BITS) + (val & LOWER_BITS_MASK);
+    }
+
+    /*!
+     * \brief returns a cumulative histogram.
+     *
+     * The metrics cumulative histogram uses upper bounds.
+     * The histogram.count serves as an infinite upper bound bucket
+     */
+    seastar::metrics::histogram get_histogram() const {
+        seastar::metrics::histogram res;
+        res.buckets.resize(NUM_BUCKETS - 1);
+        uint64_t cummulative_count = 0;
+        res.sample_sum = 0;
+
+        for (size_t i = 0; i < NUM_BUCKETS - 1; i++) {
+            auto& v = res.buckets[i];
+            v.upper_bound = get_bucket_limit(i + 1);
+            cummulative_count += buckets[i];
+            v.count = cummulative_count;
+            res.sample_sum += buckets[i] * v.upper_bound;
+        }
+        // The count serves as the infinite bucket
+        res.sample_count = cummulative_count + buckets[NUM_BUCKETS - 1];
+        res.sample_sum += buckets[NUM_BUCKETS - 1] * get_bucket_limit(NUM_BUCKETS - 1);
+        return res;
+    }
+
+    /*!
+     * \brief return the histogram buckets
+     */
+    const std::vector<uint64_t>& get_buckets() const {
+        return buckets;
+    }
+
+    void clear() {
+        std::fill(buckets.begin(), buckets.end(), 0);
+    }
+
+    /*!
+     * \brief Add an item to the histogram
+     * Increments the count of the bucket closest to n, rounding DOWN.
+     */
+    histogram_type& add(uint64_t n) {
```

Comment:
It's better not to return *this, it just muddies the interface and is usually not needed.